Validate array function callback

BackEndTea opened the pull request:
Fixes https://github.com/phpstan/phpstan/issues/3883

This adds type checks for callables on array_map, array_reduce and array_walk.

Currently `array_filter` is not checked, as the params for the callable depend on the mode param, so that may need an extension of some kind.

This alos fixes new issues found by these checks

BackEndTea commented on stubs/array.stub:
```diff
@@ -0,0 +1,55 @@
+ *
+ * @return array<TKey, TReturn>
+ */
+function array_map(
```
Looking at the php docs, this one is incorrect, and it may need an extension as well.

ondrejmirtes replied:
How is it incorrect?

BackEndTea replied:
Every extra array becomes another param of the callback. e.g.

```php
array_map(
  fn(string $one, string $two) => $one . $two,
  ['a', 'b', 'c'],
  ['d', 'e', 'f']
); // ['ad', 'be', 'cf']

BackEndTea commented on src/Parallel/ParallelAnalyser.php:
```diff
@@ -157,6 +157,7 @@ public function analyse(
 						continue;
 					}
 					$exportedNodes[$file] = array_map(static function (array $node): ExportedNode {
+						//TODO: fix error on this line
```
i'm not 100% on what the fix is, most likely the annotation a few lines up is wrong, but i'm not sure

BackEndTea commented on tests/PHPStan/Rules/Functions/data/array_map.php:
```diff
@@ -0,0 +1,6 @@
+<?php declare(strict_types = 1);
+
+array_map(
+	fn(stdClass $in): string => '',
```
Self note: tests need to run on php  7.1, so these need to use a normal closure instead of the short version

ondrejmirtes commented:
Hi, thank you for this. I pushed some commits with improvements. The reason why I postponed doing this is because most functions seem pretty complicated to be described with just generics.

ondrejmirtes commented on src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php:
```diff
@@ -49,16 +56,24 @@ public function findFunctionReflection(string $functionName): ?NativeFunctionRef
 					$type = $parameterSignature->getType();
+					$phpDocType = null;
+					if ($phpDoc !== null) {
+						$phpDocParam = $phpDoc->getParamTags()[$parameterSignature->getName()] ?? null;
```
Unfortunately this needs more complicated logic - the parameter names change across PHP versions so the stub will definitely be wrong and the PHPDoc unused. We can wait and see for the tests - I suspect that different tests will fail on different PHP versions.

ondrejmirtes replied:
A similar logic is already applied when inheriting PHPDocs implicitly - parameter names can change in child classes and PHPDocs are mapped based on that.

`StubPhpDocProvider::findFunctionPhpDoc()` would need a similar logic like `findMethodPhpDoc` already has - the `$positionalParameterNames` parameter and calling `changeParameterNamesByMapping`.

ondrejmirtes reviewed (changes requested):
To be worth finishing this and merging this, I'd like more examples of these functions that would benefit from this. Like `usort()` etc.

BackEndTea commented:
> Hi, thank you for this. I pushed some commits with improvements. The reason why I postponed doing this is because most functions seem pretty complicated to be described with just generics.

True, looking into it the more commonly used functions like array_map, need more than just generics.

Prehaps some kind of `DynamicParameterTypeExtension`? though i'm not sure how to go about implementing that, that should be something for another PR.

BackEndTea commented:
I've added the stubs for usort uasort and uksort. And removed array_map, as it is incorrect.

> `StubPhpDocProvider::findFunctionPhpDoc()` would need a similar logic like `findMethodPhpDoc` already has - the `$positionalParameterNames` parameter and calling `changeParameterNamesByMapping`.

I'm not sure how i should go about this

ondrejmirtes commented:
The way it works is:

1) The stub has parameters `$a` and `$b`. PHPStan notices that `$a` has `string` and `$b` has int.
2) The real function has parameters `$x` and `$y`. Because `$x` is the first one, the type for `$a` should be used for `$x`. Because `$y` is the second one, the type for `$b` should be used. So `$x` has `string` and `$y` has int.

How to implement it here:

1) Edit the stub so that the parameters are always different from PHP. That way you can easily test the logic.
2) Make `findFunctionPhpDoc` also work with `$positionalParameterNames` same as `findMethodPhpDoc` does it.
3) Observe how `$positionalParameterNames` are passed into the `findMethodPhpDoc` method and do the same thing for `findFunctionPhpDoc`.
4) The tests testing the stubs should now pass.

BackEndTea commented on tests/PHPStan/Rules/Functions/CallToFunctionParametersRuleTest.php:
```diff
@@ -554,4 +554,66 @@ public function testBugNumberFormatNamedArguments(): void
+				13,
+			],
+			[
+				'Parameter #2 $callback of function array_reduce expects callable(string|null, int): string|null, Closure(string, int): string given.',
```
This is the last remaining test failure. If no argument is passed, then the first function call will take null, as the `carry` param.

Not sure if we should ignore this issue for now, or drop array_reduce as well

ondrejmirtes replied:
This highlights a separate issue that's not related to generics. The problem is that we don't know the default value of parameters for built-in functions. For `array_reduce` this place should pass in `new NullType()` instead of `null`: https://github.com/phpstan/phpstan-src/blob/2abb92e2588808b3a8dd5cc3206041b7afa9891d/src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php#L103 (you can try to edit it locally and see if the test passes).

I don't think we have to block this PR on this so feel free to comment out the test and I'll think of something. We might be able to extract this information from phpstorm-stubs and https://github.com/phpstan/php-8-stubs, but that's a completely separate feature request...

BackEndTea replied:
I've fixed this in a similar matter as fputcsv, and sprintf.

BackEndTea commented:
The (extra) failures on PHP 8 are due to param names

ondrejmirtes commented:
> The (extra) failures on PHP 8 are due to param names

The tests need to account for this with `PHP_VERSION_ID >= 80000` ternary operators filling in the correct names. Using `sprintf()` might be best for this. Thanks.

BackEndTea commented:
I can't figure out why for that one job the tests would fail.

ondrejmirtes commented:
I'll look into it. Thanks!

BackEndTea commented:
Thanks for the merge and the help @ondrejmirtes!
